crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,455 @@
+/// if sys.platform == "win32":
+///     __all__ += ["y", "z"]
+/// ```
+///
+/// ## Fix safety
+/// This rule's fix should be safe for single-line `__all__` definitions
+/// and for multiline `__all__` definitions without comments.
+/// For multiline `__all__` definitions that include comments,
+/// the fix is marked as unsafe, as it can be hard to tell where the comments
+/// should be moved to when sorting the contents of `__all__`.
+#[violation]
+pub struct UnsortedDunderAll;
+
+impl Violation for UnsortedDunderAll {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`__all__` is not alphabetically sorted")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Sort `__all__` alphabetically".to_string())
+    }
+}
+
+/// RUF022
+pub(crate) fn sort_dunder_all(checker: &mut Checker, stmt: &ast::Stmt) {
+    // We're only interested in `__all__` in the global scope
+    if !checker.semantic().current_scope().kind.is_module() {
+        return;
+    }
+
+    // We're only interested in `__all__ = ...` and `__all__ += ...`
+    let (target, original_value) = match stmt {
+        ast::Stmt::Assign(ast::StmtAssign { value, targets, .. }) => match targets.as_slice() {
+            [ast::Expr::Name(ast::ExprName { id, .. })] => (id, value.as_ref()),
+            _ => return,
+        },
+        ast::Stmt::AugAssign(ast::StmtAugAssign {
+            value,
+            target,
+            op: ast::Operator::Add,
+            ..
+        }) => match target.as_ref() {
+            ast::Expr::Name(ast::ExprName { id, .. }) => (id, value.as_ref()),
+            _ => return,
+        },
+        _ => return,
+    };
+
+    if target != "__all__" {
+        return;
+    }
+
+    let locator = checker.locator();
+
+    let Some(dunder_all_val) = DunderAllValue::from_expr(original_value, locator) else {
+        return;
+    };
+
+    let new_dunder_all = match dunder_all_val.construct_sorted_all(locator, stmt, checker.stylist())
+    {
+        SortedDunderAll::AlreadySorted => return,
+        SortedDunderAll::Sorted(value) => value,
+    };
+
+    let dunder_all_range = dunder_all_val.range();
+    let mut diagnostic = Diagnostic::new(UnsortedDunderAll, dunder_all_range);
+
+    if let Some(new_dunder_all) = new_dunder_all {
+        let applicability = {
+            if dunder_all_val.multiline
+                && checker
+                    .indexer()
+                    .comment_ranges()
+                    .intersects(original_value.range())
+            {
+                Applicability::Unsafe
+            } else {
+                Applicability::Safe
+            }
+        };
+        diagnostic.set_fix(Fix::applicable_edit(
+            Edit::range_replacement(new_dunder_all, dunder_all_range),
+            applicability,
+        ));
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+struct DunderAllValue<'a> {
```

Comment:
And I'll say that it isn't obvious to me that this invariant holds, but I think, for example, `construct_sorted_all` depends on it (it will panic, I believe, if `self.items` is empty). I was thinking this invariant was upheld by `if elts.len() < 2 { return None }` below. But `elts` gets passed through `collect_dunder_all_lines` and then `collect_dunder_all_items`. And it's the result of that that is used for `DunderAllValue::items`. But is the length preserved through those operations? I'm not sure. (I mused about this below too.)